# Data Cleaning

In [1]:
#Import the libraries
import pandas as pd
import numpy as np

In [24]:
#provide context for the column
data = {
    'column':['distance', 'consume', 'speed', 'temp_inside', 'temp_outside', 'specials', 'gas_type', 'AC', 'rain', 'sun', 'refill liters', 'refill gas'],
    'description':['distance (km)', 'consumption (L/100Km)', 'speed (Km/h)', 'temperature inside (°C)', 'temperature outside (°C)', 'anything special that happened', 'gas type used', 'whether the AC was on or off', 'whether it was raining', 'whether it was sunny', 'how much gas was bought (L)', 'gas type that was used']
}

datainfo = pd.DataFrame(data)
datainfo

column                     description
0        distance                   distance (km)
1         consume           consumption (L/100Km)
2           speed                    speed (Km/h)
3     temp_inside         temperature inside (°C)
4    temp_outside        temperature outside (°C)
5        specials  anything special that happened
6        gas_type                   gas type used
7              AC    whether the AC was on or off
8            rain          whether it was raining
9             sun            whether it was sunny
10  refill liters     how much gas was bought (L)
11     refill gas          gas type that was used

In [52]:
#Load the csv dataset
df = pd.read_csv("../data/raw/measurements.csv")
df.head()

distance consume  speed temp_inside  temp_outside specials gas_type  AC  \
0       28       5     26        21,5            12      NaN      E10   0   
1       12     4,2     30        21,5            13      NaN      E10   0   
2     11,2     5,5     38        21,5            15      NaN      E10   0   
3     12,9     3,9     36        21,5            14      NaN      E10   0   
4     18,5     4,5     46        21,5            15      NaN      E10   0   

   rain  sun refill liters refill gas  
0     0    0            45        E10  
1     0    0           NaN        NaN  
2     0    0           NaN        NaN  
3     0    0           NaN        NaN  
4     0    0           NaN        NaN

In [4]:
#Load the excel dataset
df1 = pd.read_excel("../data/raw/measurements2.xlsx")
df1.head()

distance  consume  speed  temp_inside  temp_outside specials gas_type  AC  \
0      28.0      5.0     26         21.5            12      NaN      E10   0   
1      12.0      4.2     30         21.5            13      NaN      E10   0   
2      11.2      5.5     38         21.5            15      NaN      E10   0   
3      12.9      3.9     36         21.5            14      NaN      E10   0   
4      18.5      4.5     46         21.5            15      NaN      E10   0   

   rain  sun  refill liters refill gas  
0     0    0           45.0        E10  
1     0    0            NaN        NaN  
2     0    0            NaN        NaN  
3     0    0            NaN        NaN  
4     0    0            NaN        NaN

In [5]:
df.shape

(388, 12)

In [6]:
df1.shape

(388, 12)

In [7]:
df.isna().sum()

distance           0
consume            0
speed              0
temp_inside       12
temp_outside       0
specials         295
gas_type           0
AC                 0
rain               0
sun                0
refill liters    375
refill gas       375
dtype: int64

In [8]:
df1.isna().sum()

distance           0
consume            0
speed              0
temp_inside       12
temp_outside       0
specials         295
gas_type           0
AC                 0
rain               0
sun                0
refill liters    375
refill gas       375
dtype: int64

In [9]:
df.dtypes

distance         object
consume          object
speed             int64
temp_inside      object
temp_outside      int64
specials         object
gas_type         object
AC                int64
rain              int64
sun               int64
refill liters    object
refill gas       object
dtype: object

In [10]:
df1.dtypes

distance         float64
consume          float64
speed              int64
temp_inside      float64
temp_outside       int64
specials          object
gas_type          object
AC                 int64
rain               int64
sun                int64
refill liters    float64
refill gas        object
dtype: object

Based on this observation, we can see that the two datasets are the same. However, the data enclosed in the excel file seems easier to work with as values in each columns are in the correct types.

In [11]:
#Lets make a copy of the original dataset
measurement = df1.copy()

In [12]:
measurement.head()

distance  consume  speed  temp_inside  temp_outside specials gas_type  AC  \
0      28.0      5.0     26         21.5            12      NaN      E10   0   
1      12.0      4.2     30         21.5            13      NaN      E10   0   
2      11.2      5.5     38         21.5            15      NaN      E10   0   
3      12.9      3.9     36         21.5            14      NaN      E10   0   
4      18.5      4.5     46         21.5            15      NaN      E10   0   

   rain  sun  refill liters refill gas  
0     0    0           45.0        E10  
1     0    0            NaN        NaN  
2     0    0            NaN        NaN  
3     0    0            NaN        NaN  
4     0    0            NaN        NaN

In [13]:
measurement['specials'].unique()

array([nan, 'AC rain', 'AC', 'rain', 'snow', 'AC snow',
       'half rain half sun', 'sun', 'AC sun', 'sun ac', 'ac', 'AC Sun',
       'ac rain'], dtype=object)

In [14]:
#define a function to clean the columns of the dataset
def clean_cols(df):
    df.columns = df.columns.str.replace(' ', '_').str.lower()

    return df


#apply the function to the 'measurements' df
measurement = clean_cols(measurement)
measurement.head()

distance  consume  speed  temp_inside  temp_outside specials gas_type  ac  \
0      28.0      5.0     26         21.5            12      NaN      E10   0   
1      12.0      4.2     30         21.5            13      NaN      E10   0   
2      11.2      5.5     38         21.5            15      NaN      E10   0   
3      12.9      3.9     36         21.5            14      NaN      E10   0   
4      18.5      4.5     46         21.5            15      NaN      E10   0   

   rain  sun  refill_liters refill_gas  
0     0    0           45.0        E10  
1     0    0            NaN        NaN  
2     0    0            NaN        NaN  
3     0    0            NaN        NaN  
4     0    0            NaN        NaN

In [15]:
measurement.isna().sum()

distance           0
consume            0
speed              0
temp_inside       12
temp_outside       0
specials         295
gas_type           0
ac                 0
rain               0
sun                0
refill_liters    375
refill_gas       375
dtype: int64

For both 'refill_liters' and 'refill_gas', we can see that there are 375 missing values out of 388 rows. As majority of teh values in these two columns are missing, none of the fill methods would be appropriate. Likewise, the 'specials' column has 295 missing values and does not really provide additional information. Therefore, 'refill_liters', 'refill_gas', and 'specials' will be dropeed from the dataset.

In [16]:
cols_to_drop = ['refill_liters', 'refill_gas', 'specials']
measurement = measurement.drop(cols_to_drop, axis = 1)
measurement.head()

distance  consume  speed  temp_inside  temp_outside gas_type  ac  rain  sun
0      28.0      5.0     26         21.5            12      E10   0     0    0
1      12.0      4.2     30         21.5            13      E10   0     0    0
2      11.2      5.5     38         21.5            15      E10   0     0    0
3      12.9      3.9     36         21.5            14      E10   0     0    0
4      18.5      4.5     46         21.5            15      E10   0     0    0

In [17]:
measurement.isna().sum()

distance         0
consume          0
speed            0
temp_inside     12
temp_outside     0
gas_type         0
ac               0
rain             0
sun              0
dtype: int64

In [18]:
measurement[measurement['temp_inside'].isnull()]

distance  consume  speed  temp_inside  temp_outside gas_type  ac  rain  \
93       12.4      4.7     43          NaN            10     SP98   0     0   
95       11.8      5.3     52          NaN            11     SP98   0     0   
97       15.7      5.3     33          NaN             9     SP98   0     0   
98       12.9      5.7     35          NaN             9     SP98   0     0   
99        6.4      4.4     37          NaN            10     SP98   0     0   
100       5.3      4.1     34          NaN             9     SP98   0     0   
102      18.8      5.0     62          NaN             9     SP98   0     1   
201      22.2      3.8     42          NaN            15     SP98   0     0   
203      12.6      4.1     33          NaN            17     SP98   0     0   
261      24.5      3.9     50          NaN            15      E10   0     0   
267      22.9      4.0     29          NaN            21      E10   0     0   
268      16.0      3.8     42          NaN             8      E10   0     0   

     sun  
93     0  
95     0  
97     0  
98     0  
99     0  
100    0  
102    0  
201    0  
203    0  
261    1  
267    0  
268    0

Finally, the 'temp_inside' column still contains null values. As there are only 12 null values and it is numerical, we will replace null values with the mean

In [19]:
#calculate the mean of values in 'temp_inside'
mean = measurement['temp_inside'].mean()
mean

21.929521276595743

In [20]:
#round mean to 1 decimal place
rounded_mean = round(mean, 1)
rounded_mean

21.9

In [21]:
#replacing null values in 'temp_inside' with mean 
measurement['temp_inside'] = measurement['temp_inside'].fillna(rounded_mean)
measurement.head()

distance  consume  speed  temp_inside  temp_outside gas_type  ac  rain  sun
0      28.0      5.0     26         21.5            12      E10   0     0    0
1      12.0      4.2     30         21.5            13      E10   0     0    0
2      11.2      5.5     38         21.5            15      E10   0     0    0
3      12.9      3.9     36         21.5            14      E10   0     0    0
4      18.5      4.5     46         21.5            15      E10   0     0    0

In [22]:
measurement.isna().sum()

distance        0
consume         0
speed           0
temp_inside     0
temp_outside    0
gas_type        0
ac              0
rain            0
sun             0
dtype: int64

In [23]:
#Check to make sure the missing values have been correctly replaced by the mean
measurement['temp_inside'].iloc[261]

21.9

In [25]:
measurement['ac'].unique()

array([0, 1])

In [26]:
measurement['rain'].unique()

array([0, 1])

In [27]:
measurement['sun'].unique()

array([0, 1])

In [28]:
measurement['gas_type'].unique()

array(['E10', 'SP98'], dtype=object)

Here are the current average prices for the two fuel types in the Netherlands:

- E10 = €1.981
- SP98 = €2.219

In [31]:
#we can create a column of how much fuel was consumed in each trip
measurement['fuel_consumed'] = measurement['distance'] * measurement['consume']/100
measurement.head()

distance  consume  speed  temp_inside  temp_outside gas_type  ac  rain  \
0      28.0      5.0     26         21.5            12      E10   0     0   
1      12.0      4.2     30         21.5            13      E10   0     0   
2      11.2      5.5     38         21.5            15      E10   0     0   
3      12.9      3.9     36         21.5            14      E10   0     0   
4      18.5      4.5     46         21.5            15      E10   0     0   

   sun  fuel_consumed  
0    0         1.4000  
1    0         0.5040  
2    0         0.6160  
3    0         0.5031  
4    0         0.8325

In [39]:
#calculate the cost of each trip based on fuel_consumed and the gas_type
def trip_cost(df):
    if df['gas_type'] == 'E10':
        return df['fuel_consumed'] * 1.981
    else:
        return df['fuel_consumed'] * 2.219

measurement['cost_of_trip'] = measurement.apply(trip_cost, axis = 1)
measurement.head()

distance  consume  speed  temp_inside  temp_outside gas_type  ac  rain  \
0      28.0      5.0     26         21.5            12      E10   0     0   
1      12.0      4.2     30         21.5            13      E10   0     0   
2      11.2      5.5     38         21.5            15      E10   0     0   
3      12.9      3.9     36         21.5            14      E10   0     0   
4      18.5      4.5     46         21.5            15      E10   0     0   

   sun  fuel_consumed  cost_of_trip  
0    0         1.4000      2.773400  
1    0         0.5040      0.998424  
2    0         0.6160      1.220296  
3    0         0.5031      0.996641  
4    0         0.8325      1.649183

In [40]:
measurement_numerical = measurement.select_dtypes(include=['number'])
measurement_categorical = measurement.select_dtypes(include=['object'])

In [48]:
measurement_numerical.describe()

distance     consume       speed  temp_inside  temp_outside  \
count  388.000000  388.000000  388.000000   388.000000    388.000000   
mean    19.652835    4.912371   41.927835    21.928608     11.358247   
std     22.667837    1.033172   13.598524     0.994679      6.991542   
min      1.300000    3.300000   14.000000    19.000000     -5.000000   
25%     11.800000    4.300000   32.750000    21.500000      7.000000   
50%     14.600000    4.700000   40.500000    22.000000     10.000000   
75%     19.000000    5.300000   50.000000    22.500000     16.000000   
max    216.100000   12.200000   90.000000    25.500000     31.000000   

               ac        rain         sun  fuel_consumed  cost_of_trip  
count  388.000000  388.000000  388.000000     388.000000    388.000000  
mean     0.077320    0.123711    0.082474       0.935294      1.976397  
std      0.267443    0.329677    0.275441       1.115967      2.395362  
min      0.000000    0.000000    0.000000       0.120000      0.237720  
25%      0.000000    0.000000    0.000000       0.551650      1.169384  
50%      0.000000    0.000000    0.000000       0.680600      1.455664  
75%      0.000000    0.000000    0.000000       0.916500      1.915107  
max      1.000000    1.000000    1.000000      11.453300     25.414873

In [51]:
measurement.to_csv('clean_cobify.csv', index = False)